include/seqan3/alignment/pairwise/execution/alignment_executor_two_way.hpp

```diff
@@ -231,112 +241,158 @@ class alignment_executor_two_way
+        if (!is_buffer_empty())  // Not everything consumed yet.
+            return underflow_status::non_empty_buffer;
 
         if (is_eof())  // Case: reached end of resource.
-            return eof;
-
-        // Reset the get pointer.
-        setg(std::ranges::begin(buffer), std::ranges::end(buffer));
+            return underflow_status::end_of_resource;
 
-        using std::get;
+        // Reset the buckets and the bucket iterator
+        reset_buckets();
 
-        // Apply the alignment execution.
-        size_t count = 0;
-        size_t buffer_limit = in_avail();
-
-        for (size_t current_chunk_size = std::ranges::distance(*chunked_resource_it);
-             count < buffer_limit && !is_eof();
-             count += current_chunk_size, gptr += current_chunk_size, ++chunked_resource_it)
+        // Execute the algorithm (possibly asynchronous) and fill the buckets in this pre-assigned order.
+        for (bucket_end = bucket_iterator; bucket_end != bucket_vector.end() && !is_eof(); ++bucket_end, ++resource_it)
         {
-            auto && current_chunk = std::ranges::iter_move(chunked_resource_it);
-            current_chunk_size = std::ranges::distance(current_chunk);
-            assert(in_avail() >= current_chunk_size);
-
-            exec_handler.execute(kernel, std::move(current_chunk), [write_to = gptr] (auto && alignment_result) mutable
+            exec_handler.execute(kernel, *resource_it, [target_bucket_iterator = bucket_end] (auto && alignment_result)
             {
-                *write_to = std::move(alignment_result);
-                ++write_to;
+                target_bucket_iterator->push_back(std::move(alignment_result));
             });
         }
 
         exec_handler.wait();
 
-        // Update the available get positions to cover the part of the buffer that was filled.
-        setg(std::ranges::begin(buffer), std::ranges::begin(buffer) + count);
+        // Move the results iterator to the next available result. (This skips empty results of the algorithm)
+        find_next_non_empty_bucket();
+
+        if (is_buffer_empty())
+            return underflow_status::empty_buffer;
 
-        return in_avail();
+        return underflow_status::non_empty_buffer;
     }
     //!\}
 
     /*!\name Miscellaneous
      * \{
      */
 
-    /*!\brief Initialises the underlying buffer.
-     * \param[in] size The initial size of the buffer.
-     * \param[in] init_value The value used to initialise the buffer.
+    /*!\brief Whether the internal buffer is empty.
+     * \returns `true` if all elements of the internal buffer have been consumed, otherwise `false`.
      */
-    void init_buffer(size_t const size, value_t const init_value)
+    bool is_buffer_empty() const
     {
-        buffer.resize(size, std::move(init_value));
-        setg(std::ranges::end(buffer), std::ranges::end(buffer));
+        return bucket_iterator == bucket_end || buffer_iterator == bucket_iterator->end();
```

Comment:
Does the second condition ever evaluate to true? Because in `next_buffer_iterator` you always increase the bucket_iterator if you are at the end of a buffer vector.

Reply:
good point. Probably not. I'll remove it and see if something breaks.